src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -241,79 +247,76 @@ rec_hash(struct d_hash_table *htable, d_list_t *rlink)
 {
 	struct dir_hdl *hdl = hdl_obj(rlink);
 
-	return d_hash_string_u32(hdl->name, strnlen(hdl->name, DFS_MAX_PATH));
+	return d_hash_string_u32(hdl->name, hdl->name_size);
 }
 
 static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
 					  .hop_rec_decref = rec_decref,
 					  .hop_rec_free   = rec_free,
 					  .hop_rec_hash   = rec_hash};
 
-/**
- * Look up a path in hash table to get the dfs_obj for a dir.
- * If not found in hash table, call to dfs_lookup() to open the object on DAOS.
- * If the object is corresponding to a dir, insert the object into hash table.
- * Since many DFS APIs need parent dir DFS object as parameter, we use a hash
- * table to cache parent objects for efficiency.
+/*
+ * Look up a path in hash table to get the dfs_obj for a dir. If not found in hash table, call to
+ * dfs_lookup() to open the object on DAOS. If the object is corresponding is a dir, insert the
+ * object into hash table. Since many DFS APIs need parent dir DFS object as parameter, we use a
+ * hash table to cache parent objects for efficiency.
+ *
+ * Places no restrictions on the length of the string which does not need to be \0 terminated. len
+ * should be length of the string, not including any \0 termination.
  */
 static int
-lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode, dfs_obj_t **obj)
+lookup_insert_dir(struct dfs_mt *mt, const char *name, size_t len, mode_t *mode, dfs_obj_t **obj)
 {
-	struct dir_hdl *hdl;
+	struct dir_hdl *hdl = NULL;
 	dfs_obj_t      *oh;
 	d_list_t       *rlink;
-	size_t          len;
 	int             rc;
 
-	if (obj == NULL) {
-		D_ERROR("Invalid pointer!");
-		return EFAULT;
-	}
-	*obj = NULL;
-	len = strnlen(name, DFS_MAX_PATH);
-	if (len >= DFS_MAX_PATH) {
-		D_ERROR("path is too long!");
-		return ENAMETOOLONG;
-	}
-	D_RWLOCK_RDLOCK(&dfs_list[idx_dfs].lock_ht);
-	rlink = d_hash_rec_find(dfs_list[idx_dfs].dfs_dir_hash, name, len);
-	D_RWLOCK_UNLOCK(&dfs_list[idx_dfs].lock_ht);
+	/* TODO: Remove this after testing. */
+	D_ASSERT(strlen(name) == len);
+
+	*obj  = NULL;
+	rlink = d_hash_rec_find(mt->dfs_dir_hash, name, len);
 	if (rlink != NULL) {
-		hdl = hdl_obj(rlink);
+		hdl  = hdl_obj(rlink);
 		*obj = hdl->oh;
 		return 0;
 	}
 
-	rc = dfs_lookup(dfs_list[idx_dfs].dfs, name, O_RDWR, &oh, mode, NULL);
+	rc = dfs_lookup(mt->dfs, name, O_RDWR, &oh, mode, NULL);
 	if (rc)
 		return rc;
 
-	if (mode && !S_ISDIR(*mode)) {
+	if (!S_ISDIR(*mode)) {
 		*obj = oh;
 		return 0;
 	}
 
-	D_ALLOC_PTR(hdl);
+	/* Allocate struct and string in a single buffer.  This includes a extra byte so name will
+	 * be \0 terminiated however that is not required.
+	 */
+	D_ALLOC(hdl, sizeof(*hdl) + len + 1);
 	if (hdl == NULL) {
+		D_GOTO(out_release, rc = ENOMEM);
```

Comment:
Need to call "dfs_release(oh);" before "D_GOTO(out_release, rc = ENOMEM);".

Reply:
Thank you.